### ISSUE TITLE
Possible bug in TripletLoss- Sampling:hard 

### ISSUE THREAD

**Issue by bml1g12** [pyannote/pyannote-audio#112]
Using pyannote.audio==1.0.1

Training on AMI using the following features:
```
feature_extraction: 
 name: YaafeMFCC 
 params: 
 duration: 0.025 
 step: 0.010 
 coefs: 19 
 D: True 
 DD: True 
 e: False 
 De: True 
 DDe: True
 ```
Using the following config.yml

```
# train the network using triplet loss 
# see pyannote.audio.embedding.approaches for more details 
approach: 
 name: TripletLoss 
 params: 
 metric: cosine # embeddings are optimized for cosine metric 
 clamp: positive # triplet loss variant 
 margin: 0.2 # triplet loss margin 
 duration: 2 # sequence duration 
 per_fold: 40 # number of speakers per fold 
 per_label: 3 # number of sequences per speaker 

# use precomputed features (see feature extraction tutorial) 
feature_extraction: 
 name: Precomputed 
 params: 
 root_dir: /data/ben_scratch/pyannote_runs/AMI/feature_extractionF2/features_F2 

# see pyannote.audio.embedding.models for more details 
architecture: 
 name: ClopiNet 
 params: 
 rnn: LSTM 
 recurrent: [800] 
 bidirectional: False 
 linear: [200] 
 batch_normalize: True 
 normalize: sphere 

# use cyclic learning rate scheduler 
scheduler: 
 name: CyclicScheduler 
 params: 
 learning_rate: auto
```

Works fine 

![image](https://user-images.githubusercontent.com/2605767/43010394-50a5938e-8c7b-11e8-8be4-35831ac83adf.png)

But adding `sampling: hard` to the TripletLoss approach results in the following:

![image](https://user-images.githubusercontent.com/2605767/43010440-6ee8cbcc-8c7b-11e8-8456-d220ff3284d9.png)
![image](https://user-images.githubusercontent.com/2605767/43010485-970db3a6-8c7b-11e8-8870-05f9c60dc965.png)

I'm not sure if this is a bug or expected for this sampling strategy, but it seems suspicious that the training EER decreases whilst the validation EER is erratic. 

Also I would have expected regardless of the sampling strategy to get the same a similar converged training EER, but with a good sampling strategy to reach convergence faster. Not sure if this intuition is correct though.

**Comment by hbredin** [comment]
I implemented several sampling strategies but did not really test them thoroughly. 
I would like to do more tests in the future (e.g. starting with `sampling: all` then switching to `sampling: hard` when the model has reached a reasonable state...) but haven't found the time to do so...

I would also suggest reading [In Defense of the Triplet Loss for Person Re-Identification](https://arxiv.org/pdf/1703.07737.pdf) that tries to provide intuition as to what those various startegies actually do (plus many more interesting findings...)
